pkg/batchjob/stat.go

```diff
@@ -0,0 +1,226 @@
+		stat.TotalSeconds = int64(totalSeconds)
+	} else {
+		//if there are no hbs in this range
+		stat.Projects = []models.Workunit{}
+		stat.Languages = []models.Workunit{}
+		stat.DaysMinusHolidays = 0
+		stat.Holidays = stat.DaysIncludingHolidays
+		stat.BestDay.Date = beginTime
+		stat.BestDay.TotalSeconds = 0
+		stat.Start = beginTime.Unix()
+		stat.End = beginTime.Add(ToDuration(statRange)).Unix()
+		stat.TotalSeconds = 0
+	}
+
+	// part 4: metadata
+	stat.Range = statRange
+	stat.CreatedAt = time.Now()
+	stat.ModifiedAt = stat.CreatedAt
+
+	return &stat, nil
+}
+
+// 3 cases
+// 1. start + jiffy < old endtime -> this hb contributes nothing
+// 2. start < old endtime < time + jiffy -> this hb contributes less than a jiffy
+// 3. old endtime < start -> there is a discontinuity of work, and hb contributes a jiffy
+func mergeIntervals(prevEnd float64, start float64, jiffy float64) (elapsed float64, newEnd float64) {
+	if start+jiffy < prevEnd {
+		return 0, prevEnd
+	} else if start < prevEnd {
+		return start + jiffy - prevEnd, start + jiffy
+	} else {
+		return jiffy, start + jiffy
+	}
+}
+
+func upsertWorkunit(workunits *Workunits, name string, start float64, jiffy float64) {
+	/* if such name does not exist, create one ending in 1970 */
+	if _, ok := workunits.units[name]; !ok {
+		workunits.units[name] = WorkunitWithEndtime{
+			Workunit: models.Workunit{
+				Name:         name,
+				Percent:      0,
+				TotalSeconds: 0,
+			},
+			Endtime: 0,
+		}
+	}
+
+	/* update it */
+	wu := workunits.units[name]
+	elapsed, newEnd := mergeIntervals(wu.Endtime, start, jiffy)
+	wu.Endtime = newEnd
+	wu.TotalSeconds += int64(elapsed)
+	workunits.totalSeconds += elapsed
+}
+
+//WorkunitWithEndtime is a temporal struct for aggregation
+type WorkunitWithEndtime struct {
+	models.Workunit
+	Endtime float64
+}
+
+//calculate percents for each wu, and return appropriate type for output.
+func percentedWorkunits(wus Workunits) []models.Workunit {
+	outputs := make([]models.Workunit, len(wus.units))
+	for _, wu := range wus.units {
+		wu.Percent = float64(wu.TotalSeconds) / wus.totalSeconds
+		outputs = append(outputs, wu.Workunit)
+	}
+	return outputs
+}
+
+//Workunits is a temporal struct for aggregation
+type Workunits struct {
+	units        map[string]WorkunitWithEndtime
+	totalSeconds float64
+}
+
+//ToDuration converts semantics of statRange into time.Duration.
+//FIXME: it is an approximation.
+//To get real duration, we have to decide whether use calendar date or duration date
+//and have to have another argument of beginTime.
+func ToDuration(statRange models.StatRange) time.Duration {
+	switch statRange {
+	case models.StatRangeLast7Days:
+		return time.Hour * 24 * 7
+	case models.StatRangeLast30Days:
+		return time.Hour * 24 * 30
+	case models.StatRangeLast6Months:
+		return time.Hour * 24 * 30 * 6
+	case models.StatRangeLastYear:
+		return time.Hour * 24 * 365
+	}
+	// will not happen
+	log.Fatalln("unexpected case of statRange")
```

Comment:
you can  use panic, this will indicate to the compiler you want to stop execution (I don't think it will follow log.Fatal and found it calls os.Exit), also it will print a stack trace when not recovered.

Also I'd suggest put this into the `default` in `switch`